src/_guides/language/effective-dart/design_migrated.md

```diff
@@ -933,20 +938,31 @@ not intended to be invoked from Dart code and don't need a corresponding getter.
 [angular]: {{site.angulardart}}
 
 
-### AVOID returning `null` from members whose return type is `bool`, `double`, `int`, or `num`.
+### AVOID public `late final` fields.
+
+Unlike other final fields, a late final field *does* define a setter. If
+that field is public, then the setter is public. This is rarely what you want.
+Fields are usually marked late so that they can be initialized *internally* at
+some point in the instance's lifetime, often inside the constructor body.
+
+Unless you do intend to expose the setter publicly and allow users to invoke it,
+it's better to either not use `late`, or make the late field private with a
+public getter.
+
 
-{% include linter-rule.html rule="avoid_returning_null" %}
+### AVOID returning nullable `Future`, `Stream`, and collection types.
 
-Even though all types are nullable in Dart, users assume those types almost
-never contain `null`, and the lowercase names encourage a "Java primitive"
-mindset.
+When an API returns a container type, it has two ways to indicate the absence of
+data: It can return an empty container or it can return `null`. Users generally
+assume and prefer that you use an empty container to indicate "no data". That
+way, they have a real object that they can call methods on like `isEmpty`.
 
-It can be occasionally useful to have a "nullable primitive" type in your API,
-for example to indicate the absence of a value for some key in a map, but these
-should be rare.
+Prefer returning an empty collection, a future that doesn't complete, or a
```

Comment:
A future that doesn't complete is not often what users would want. More likely you want a non-nullable future with a nullable generic.

Reply:
Ah, good call. Done.